extras/gizmo/tri.js

```diff
@@ -0,0 +1,73 @@
+import {
+    Vec3
+} from 'playcanvas';
+
+const e1 = new Vec3();
+const e2 = new Vec3();
+const h = new Vec3();
+const s = new Vec3();
+const q = new Vec3();
+
+// constants
+const EPSILON = 1e-6;
+
+class Tri {
+    v0 = new Vec3();
+
+    v1 = new Vec3();
+
+    v2 = new Vec3();
+
+    constructor(v0 = new Vec3(), v1 = new Vec3(), v2 = new Vec3()) {
+        if (v0.length === 3) {
+            this.v0.copy(v0[0]);
+            this.v1.copy(v0[1]);
+            this.v2.copy(v0[2]);
+        } else {
+            this.v0.copy(v0);
+            this.v1.copy(v1);
+            this.v2.copy(v2);
+        }
+    }
+
+    set(v0, v1, v2) {
+        this.v0.copy(v0);
+        this.v1.copy(v1);
+        this.v2.copy(v2);
+
+        return this;
+    }
+
+    intersectRay(origin, dir, out = new Vec3(), epsilon = EPSILON) {
```

Comment:
I would not default `out` to `new Vec3` - the user cannot get the result in that case anyways. It can just be left undefined, and test for it when writing to it.

Alternatively, leave the signature as is, but return the Vec3 (with data) in case of intersection, and null if no intersection.